### PR TITLE
Use `istiod-` as the prefix for the DNS names for Istio revisions

### PR DESCRIPTION
**Pull request by wallrj**
Fixes: #432 

Istio CSR worked fine on OpenShift when I installed it using a basic commands which are now documented here:
 * https://docs.venafi.cloud/vaas/k8s-components/t-istiocsr-vcpo-install/#step-6-install-istio

But I encountered a problem with Istio CSR when I tried to get it working with OpenShift Open Service Mesh.

> 2024-12-10T17:07:29.843926Z     warning envoy config external/envoy/source/extensions/config_subscription/grpc/grpc_stream.h:190        StreamAggregatedResources gRPC config stream to xds-grpc closed since 15485s ago: 14, connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate is valid for istiod.istio-system.svc, istiodbasic.istio-system.svc, not istiod-basic.istio-system.svc"   thread=19

TBH I don't understand all the terminology and components in Istio, but here's a summary of how I triggered the bug.

I used the following resource to install Istio with the RedHat Open Service Mesh Operator:

```yaml
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  version: v2.6
  tracing:
    type: None
    sampling: 10000
  addons:
    kiali:
      enabled: true
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.venafi.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
```
And it resulted in Istio being deployed with a "revision" called "basic".
I added that "revision" to the Istio CSR config and it resulted in an Istio serving certificate  with an incorrect DNS name which caused the istio-ingressgateway to log an error and not become ready:

With the change in this branch, the istio-ingressgateway becomes ready and the error is no longer logged.

## Testing

I published the Docker images and Helm chart to a temporary registry as follows:

```bash
> release.env
make release \
      OCI_SIGN_ON_PUSH=false \
      oci_platforms=linux/amd64 \
      oci_manager_image_name=$OCI_BASE/images/cert-manager-istio-csr \
      helm_chart_repo_base=oci://$OCI_BASE/charts \
      GITHUB_OUTPUT=release.env
source release.env
helm push $RELEASE_HELM_CHART_TAR oci://$OCI_BASE/charts
```

Applied an VenafiInstall for use in an OpenShift cluster where the Venafi Control Plane Operator is already running:
```yaml
# venafi-components.yaml
apiVersion: installer.venafi.com/v1alpha1
kind: VenafiInstall
metadata:
  name: venafi-components
spec:
  globals:
    enableDefaultApprover: true # Istio CSR works best with the cert-manager default approver.
    imagePullSecretNames: [venafi-image-pull-secret]
    namespace: venafi
  certManager:
    install: true
  certManagerIstioCSR:
    install: true
    version: v0.13.0-11-g9be4830851493f
    imageRegistry: ttl.sh/81b08360-c679-425b-ab37-82591b0ee347/images
    chartRepository: oci://ttl.sh/81b08360-c679-425b-ab37-82591b0ee347/charts
    trustDomain: cluster.local
    runtimeConfigMapName: istio-csr-ca
    values:
      app:
        istio:
          namespace: istio-system
          revisions: ["default", "basic"]
```

```bash
oc apply -f venafi-components.yaml
```

Previously the istio-ingressgateway had been failing, then after updating Istio CSR the Readiness check succeeded and the deployment became ready:

```console
[crc@crc-4-17 ~]$ oc -n istio-system logs deploy/istio-ingressgateway
...
2024-12-10T17:07:29.843926Z     warning envoy config external/envoy/source/extensions/config_subscription/grpc/grpc_stream.h:190        StreamAggregatedResources gRPC config stream to xds-grpc closed since 15485s ago: 14, connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate is valid for istiod.istio-system.svc, istiodbasic.istio-system.svc, not istiod-basic.istio-system.svc"   thread=19
2024-12-10T17:07:38.002565Z     error   failed scraping envoy metrics: error scraping http://localhost:15090/stats/prometheus: Get "http://localhost:15090/stats/prometheus": dial tcp [::1]:15090: connect: connection refused
2024-12-10T17:07:53.001930Z     error   failed scraping envoy metrics: error scraping http://localhost:15090/stats/prometheus: Get "http://localhost:15090/stats/prometheus": dial tcp [::1]:15090: connect: connection refused
2024-12-10T17:07:53.023725Z     info    xdsproxy        connected to upstream XDS server: istiod-basic.istio-system.svc:15012
2024-12-10T17:07:53.041693Z     info    ads     ADS: new connection for node:istio-ingressgateway-948d89b64-r2f99.istio-system-1
2024-12-10T17:07:53.041857Z     info    ads     ADS: new connection for node:istio-ingressgateway-948d89b64-r2f99.istio-system-2
2024-12-10T17:07:54.368463Z     info    cache   generated new workload certificate      latency=1.326661696s ttl=59m58.631542991s
2024-12-10T17:07:54.368657Z     info    cache   returned workload certificate from cache        ttl=59m58.631350077s
2024-12-10T17:07:54.368998Z     info    ads     SDS: PUSH request for node:istio-ingressgateway-948d89b64-r2f99.istio-system resources:1 size:3.2kB resource:default
2024-12-10T17:07:54.369109Z     info    ads     SDS: PUSH request for node:istio-ingressgateway-948d89b64-r2f99.istio-system resources:1 size:562B resource:ROOTCA
```

### CODE REVIEW

File: pkg/istiodcert/worker.go [hunk 1]
```diff
@@ -261,7 +272,7 @@ func makeDNSNamesFromRevisions(namespace string, istioRevisions []string) (strin
 			continue
 		}
 
-		dnsNames = append(dnsNames, fmt.Sprintf("istiod%s.%s.svc", revision, namespace))
+		dnsNames = append(dnsNames, fmt.Sprintf("istiod-%s.%s.svc", revision, namespace))
```
Comment (wallrj) [inline pkg/istiodcert/worker.go:275]: Here's an example of Istio using this prefix. It's the only example I could find in their code. There are various other examples in their test code and YAML:
 * https://github.com/istio/istio/blob/aa777b26ee935bb7e3b943a4925b76ea7320494a/istioctl/pkg/multixds/gather.go#L272-L277
 *  https://github.com/search?q=repo%3Aistio%2Fistio+%22istiod-%25s%22&type=code

File: pkg/istiodcert/worker_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			expectedRevisionPrefix := "istiod-"
+			expectedSuffix := fmt.Sprintf(".%s.svc", tc.namespace)
+			expectedCommonName := "istiod" + expectedSuffix
+
+			commonName, dnsNames := makeDNSNamesFromRevisions(tc.namespace, tc.revisions)
+
+			assert.Equal(t, expectedCommonName, commonName,
+				"The commonName should always be istiod.<namespace>.svc")
+
+			foundRevisions := make([]string, len(dnsNames))
```
Comment (inteon) [inline pkg/istiodcert/worker_test.go:75]: I would have preferred listing the expected dnsNames explicitly for each testcase and matching them.
But your solution works too.

**Comment by inteon** [comment]
The DNS names in go code now match the DNS names in YAML and what Istio is expecting https://github.com/cert-manager/istio-csr/pull/454#discussion_r1878544110. Thank you for finding and fixing the bug.
/approve
/lgtm

**Pull request merged**